source/sigv4.c

```diff
@@ -794,6 +899,332 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+            {
+                ret = true;
+            }
+            /* It is a leading space if no characters have been written yet. */
+            else if( trimmedLength == 0U )
+            {
+                ret = true;
+            }
+        }
+
+        return ret;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t writeSignedHeaderToCanonicalRequest( size_t headerIndex,
+                                                              uint32_t flags,
+                                                              CanonicalContext_t * canonicalRequest )
+    {
+        char * pBufLoc;
+        size_t buffRemaining, keyLen = 0, i = 0, trimKeyLen = 0;
+        const char * headerKey;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerIndex ].key.dataLen;
+
+        headerKey = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData;
+
+        for( i = 0; i < keyLen; i++ )
+        {
+            /* If the header field is not in canonical form already, we need to check
+             * whether this character represents a trimmable space. */
+            if( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) &&
+                ( isTrimmableSpace( headerKey, i, keyLen, curNumOfCopiedBytes ) ) )
+            {
+                /* Cannot copy trimmable space into canonical request buffer. */
+            }
+            /* Remaining buffer space should at least accommodate the character to copy and the trailing ";" */
+            else if( buffRemaining <= 1 )
+            {
+                sigV4Status = SigV4InsufficientMemory;
+            }
+            else
+            {
+                *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerIndex ].key.pData[ i ] );
+                pBufLoc++;
+                buffRemaining = -1;
+                curNumOfCopiedBytes++;
+            }
+        }
+
+        /* Add the ending ";" character.
+         * Note: Space for ending ";" was accounted for while copying header field data to
+         * canonical request buffer. */
+        if( sigV4Status == SigV4Success )
+        {
+            assert( buffRemaining >= 1 );
+            *pBufLoc = ';';
+            pBufLoc++;
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = ( buffRemaining - 1 );
+        }
+
+        return sigV4Status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, keyLen = 0, i = 0;
```

Comment:
`keyLen` and `i` are unused
```suggestion
        size_t noOfHeaders = 0;
```